rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,133 @@
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/serializer/serializer.go#L315-L338) (it ultimately lands
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/sketch_series.go#L103-L269)). Public `.proto`
+  definition can be found
+  [here](https://github.com/DataDog/agent-payload/blob/master/proto/metrics/agent_payload.proto#L47-L81).
+
+Vector has a nice description of its [metrics data
+model](https://vector.dev/docs/about/under-the-hood/architecture/data-model/metric/) and a [concise enum for
+representing it](https://github.com/timberio/vector/blob/master/lib/vector-core/src/event/metric.rs#L135-L169).
+
+
+The implementation would the consist in:
+
+- Add route for metrics in the existing `datadog_agent` source (based on both the offical
+  [API](https://docs.datadoghq.com/api/latest/metrics/) and the [Datadog Agent
+  itself](https://github.com/DataDog/datadog-agent/blob/main/pkg/forwarder/telemetry.go#L20-L31)) to cover every metric
+  type handled by this endpoint (count, gauge and rate) and
+    - Add support for missing fields in the `datadog_metrics` sinks
+    - Align tag manipulation (Datadog allows `key:foo` & `key:bar`, and the tags are encoded as a list whereas Vector
+      encode those as a map).
+    - Overall this is fairly straighforward except for the tagging issue
+- Add another route in the `datadog_agent` source to support sketches/distribution encoded using protobuf
+   - Revamp the distribution metrics in the `datadog_metrics` sink to use sketches and the associated endpoint will
+     probably be a good idea.
+   - The sketches the agent ships is based on this [paper](http://www.vldb.org/pvldb/vol12/p2195-masson.pdf) whereas
+     Vector uses what's called a summary inside the Agent, implementing the complete DDSketch support in Vector is
+     probably a good idea as sketches have convenient properties for wide consistent aggregation and limited error. To
+     support smooth migration, full DDsktech support is mandatory, as customers that emit distribution metric from
+     Datadog Agent would need it to migrate to Vector aggegation.
+
+**Regarding the tagging issue:** A -possibly temporary- work-around would be to store incoming tags with the complete "key:value" string as the key and an empty value to store those in the extisting map Vector uses to store [tags](https://github.com/timberio/vector/blob/master/lib/vector-core/src/event/metric.rs#L60) and slightly rework the `datadog_metrics` sink not to append `:` if a tag key has the empty string as the corresponding value.
+
+## Rationale
+
+- Smoother Vector integration with Datadog.
+- Needed for Vector to act as a complete Datadog Agent aggregator (but further work will still be required).
+- Extend the Vector ecosystem, bring additional feature for distribution metrics that would enable consistent
+  aggregation.
+
+## Drawbacks
+
+- It arbitrary imposes internal sketches representation to follow the DDSketch paper.
```

Comment:
Hardly consider this a drawback given our existing desire to do just this 😄

Reply:
I think this could be discussed lated, I think we should use ddsketches, but we should probably not close the door to integrate other sketches later.

Reply:
I think DDSketch is the right choice if we're going to natively support a sketch within Vector, as a comment separate from wanting to support _ingesting_ sketches in their native, serialized format.

Reply:
It sounds reasonnable to me, thus I suggest to set as a prerequisite, for ingesting sketches, that Vector must get internal support for sketches (based on ddsketch), WDYT ? @tobz